xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -340,111 +328,90 @@ public boolean isEquivalentTo(RoundRobinPicker picker) {
+      int numWeightedChannels = 0;
+      double sumWeight = 0;
+      float maxWeight = 0;
+      int meanWeight = 0;
+      for (float weight : weights) {
+        if (weight > 0.0001) {
+          sumWeight += weight;
+          maxWeight = Math.max(weight, maxWeight);  
+          numWeightedChannels++;
+        }
+      }
 
-    /**
-     * Use the item's deadline as the order in the priority queue. If the deadlines are the same,
-     * use the index. Index should be unique.
-     */
-    EdfScheduler(int initialCapacity, Random random) {
-      this.prioQueue = new PriorityQueue<ObjectState>(initialCapacity, (o1, o2) -> {
-        if (o1.deadline == o2.deadline) {
-          return Integer.compare(o1.index, o2.index);
+      double scalingFactor = K_MAX_WEIGHT / maxWeight;
+      if (numWeightedChannels > 0) {
+        meanWeight = (int) Math.round(scalingFactor * sumWeight / numWeightedChannels);
+      } else {
+        meanWeight = 1;
+      }
+
+      // scales weights s.t. max(weights) == K_MAX_WEIGHT, meanWeight is scaled accordingly
+      int[] scaledWeights = new int[numChannels];
+      for (int i = 0; i < numChannels; i++) {
+        if (weights[i] < 0.0001) {
+          scaledWeights[i] = meanWeight;
         } else {
-          return Double.compare(o1.deadline, o2.deadline);
+          scaledWeights[i] = (int) Math.round(weights[i] * scalingFactor);
         }
-      });
+      }
+
+      this.scaledWeights = scaledWeights;
+      this.sizeDivisor = numChannels;
       this.random = random;
+      this.sequence = new AtomicInteger((int) (this.random.nextDouble() * UINT32_MAX));
+      
     }
 
-    /**
-     * Adds the item in the scheduler. This is not thread safe.
-     *
-     * @param index The field {@link ObjectState#index} to be added
-     * @param weight positive weight for the added object
-     */
-    void add(int index, double weight) {
-      checkArgument(weight > 0.0, "Weights need to be positive.");
-      ObjectState state = new ObjectState(Math.max(weight, MINIMUM_WEIGHT), index);
-      // Randomize the initial deadline.
-      state.deadline = random.nextDouble() * (1 / state.weight);
-      prioQueue.add(state);
+    /** Returns the next sequence number and increases sequence with wraparound. */
+    private long nextSequence() {
+      return Integer.toUnsignedLong(sequence.getAndIncrement());
     }
+  
 
-    /**
-     * Picks the next WRR object.
-     */
+    /** Selects index of next backend server. */
     int pick() {
-      synchronized (lock) {
-        ObjectState minObject = prioQueue.remove();
-        minObject.deadline += 1.0 / minObject.weight;
-        prioQueue.add(minObject);
-        return minObject.index;
+      while (true) {
+        long sequence = this.nextSequence();
+        long backendIndex = sequence % this.sizeDivisor;
+        long generation = sequence / this.sizeDivisor;
+        long weight = this.scaledWeights[(int) backendIndex];
```

Comment:
this would potentially throw index out of bound, add verification in constructor.

Reply:
The value of the backendIndex is bounded by the sizeDivisor s.t. `backendIndex: [0,sizeDivisor)`, where `sizeDivisor == scaledWeights.length`. Thus, it is guaranteed to be within the bounds of the array. However, maybe it would be more appropriate to cast to an `int` earlier and make the backendIndex an `int` altogether? What do you think? This is of course under the assumption that we have less than 0xFFFF_FFFF backends.

Reply:
You are right, the size is an integer and the algorithm guarantees it will find an index in the array.
I think it is safe to cast it earlier to int, because it won't loose any precision. The max possible size is 0x7FFF_FFFFL (Integer.MAX_VALUE).